packages/ckb-indexer/src/ckbIndexerFilter.ts

```diff
@@ -0,0 +1,341 @@
+  if (queryLock) {
+    if (instanceOfScriptWrapper(queryLock)) {
+      searchKeyLock = queryLock.script;
+    } else {
+      searchKeyLock = queryLock;
+    }
+  }
+  if (queryType && queryType !== "empty") {
+    if (instanceOfScriptWrapper(queryType)) {
+      searchKeyType = queryType.script;
+    } else {
+      searchKeyType = queryType;
+    }
+  }
+
+  if (searchKeyLock) {
+    searchKey = {
+      script: searchKeyLock,
+      scriptType: "lock",
+      scriptSearchMode: instanceOfScriptWrapper(queryLock)
+        ? queryLock.searchMode
+        : "prefix",
+      filter: {
+        script: searchKeyType,
+      },
+    };
+  } else if (searchKeyType) {
+    searchKey = {
+      script: searchKeyType,
+      scriptType: "type",
+      scriptSearchMode: instanceOfScriptWrapper(queryLock)
+        ? queryLock.searchMode
+        : "prefix",
+      filter: {
+        script: searchKeyLock,
+      },
+    };
+  } else {
+    throw new Error("query.lock and query.type can't be both empty");
+  }
+
+  if (queryOptions.fromBlock || queryOptions.toBlock) {
+    searchKey.filter!.blockRange = [
+      queryOptions.fromBlock || "0x0",
+      queryOptions.toBlock
+        ? BI.from(queryOptions.toBlock).add(1).toHexString()
+        : "0x" + Number.MAX_SAFE_INTEGER.toString(16),
+    ];
+  }
+
+  return searchKey;
+}
+
+function filterByQueryOptions(cells: Cell[], options: QueryOptions): Cell[] {
+  const searchKey = convertQueryOptionToSearchKey(options);
+  let filteredCells = cells.filter((cell) => filterBy(cell, searchKey));
+
+  if (options.argsLen && options.argsLen !== "any") {
+    filteredCells = filteredCells.filter(
+      (cell) =>
+        bytes.bytify(cell.cellOutput.lock.args).length === options.argsLen
+    );
+  }
+
+  if (!!options.data && options.data !== "any") {
+    if (
+      instanceOfDataWithSearchMode(options.data) &&
+      options.data.searchMode === "exact"
+    ) {
+      const dataSearch = options.data as DataWithSearchMode;
+      filteredCells = filteredCells.filter((cell) =>
+        bytes.equal(bytes.bytify(cell.data), bytes.bytify(dataSearch.data))
+      );
+    } else if (
+      instanceOfDataWithSearchMode(options.data) &&
+      options.data.searchMode === "prefix"
+    ) {
+      const dataSearch = options.data as DataWithSearchMode;
+      filteredCells = filteredCells.filter(
+        (cell) =>
+          Buffer.from(bytes.bytify(cell.data)).indexOf(
+            bytes.bytify(dataSearch.data)
+          ) === 0
+      );
+    } else {
+      filteredCells = filteredCells.filter(
+        (cell) =>
+          Buffer.from(bytes.bytify(cell.data)).indexOf(
+            bytes.bytify(options.data as string)
+          ) === 0
+      );
+    }
+  }
+
+  if (options.order === "desc") {
+    filteredCells.reverse();
+  }
+
+  if (options.skip) {
+    filteredCells = filteredCells.slice(options.skip);
+  }
+
+  return filteredCells;
+}
+
+function filterBy(cell: Cell, searchKey: SearchKey): boolean {
```

Comment:
Mark it as `@internal` or remove it from export list, because we should avoid functions being used incorrectly by developers

Reply:
Is it ok to just not re-export it in `index.ts`, not exporting this function may lead to inconvenience in testing

Reply:
`@internal` will only just strip the type declaration from compiled `.d.ts`, it does not affect the test

```ts
/**
 * @internal
 */
functoin filterBy() {

```

Reply:
updated